### PR TITLE
Virtual threads

### PR DESCRIPTION
**Pull request by gortiz**
This PR adds the ability to run multi-stage stages in different executors. Two executors are provided:
- A cached thread executor, which mimics the current behavior
- A virtual thread executor, which uses virtual threads, introduced in Java 21.

Changes in the code are pretty simple. The main problem here is that we need to:
- Provide code that only compiles and is seen in Java 21.
- Still support Java ~21~ 11.

Changes in the pipeline are not yet complete and that is why I open this PR as a draft

### CODE REVIEW

**Comment by hpvd** [comment]
+1
would be really interesting to see the Performance Impact of this.
E.g. for mariadb its pretty huge:
https://mariadb.com/resources/blog/benchmark-jdbc-connectors-and-java-21-virtual-threads/

**Comment by gortiz** [comment]
> would be really interesting to see the Performance Impact of this.

I don't expect an actual performance gain. In fact we tried virtual threads last summer and they were a bit less efficient than normal threads. That is expected and acceptable because in ideal situations these threads will not do any IO but mostly wait for other threads and heavy computation. The later is the worst case for virtual threads.

But given the current multi-stage topology, the easiest way to ensure tasks not get blocked is to use cached thread pools, which means in case of large QPS we end up using too many threads and even worse, in constrained environments (ie CPU limits un k8s/docker) we can consume all the CPU quota with multi-stage threads, starving the rest of the system.

**Comment by gortiz** [comment]
This PR has been closed in favor of the simpler #13921, which includes the executor-spi but not the virtual thread executor. Although it is trivial to add this executor, it has been proven difficult to add that contribution in the current Pinot CI and we don't have time to invest there right now.

StarTree is going to do some internal benchmarks using this API and in case virtual threads are good enough, we will try to contribute it into this repository.

File: pom.xml [hunk 1]
```diff
@@ -381,6 +381,66 @@
...
+                    </path>
+                    <path>
+                      <groupId>org.apache.logging.log4j</groupId>
+                      <artifactId>log4j-core</artifactId>
+                    </path>
+                    <path>
+                      <groupId>com.google.auto.service</groupId>
+                      <artifactId>auto-service</artifactId>
+                    </path>
+                  </annotationProcessorPaths>
```
Comment (gortiz) [inline pom.xml:436]: This is a change we should add in Pinot even if this is never merged

Comment (Jackie-Jiang) [reply pom.xml:436]: Added in #14185

**Pull request closed without merging**